Fix/caldav/eventcomparisionservice uses wrong array comparison

rcwschaller opened the pull request:
* Resolves: Partially nextcloud/server#41084 

## Summary
Fixes the comparison of two arrays that contain relevant differences between two events.

Old comparison implementation compares each element of the array against each other with no respect for the associated array label, which leads to wrongful removals because one value is accidentally present in a completely different label. New comparison works 'by-label' individually.
Refer to php manual https://www.php.net/manual/en/function.array-diff.php

Partly fixes nextcloud/server#41084 because changes between 'SEQUENCE' not present, 'SEQUENCE:0' and 'SEQUENCE:1' were not detected in the old implementation and thus no email update sent.

@miaulalala please inspect. Small fix for your otherwise comprehensive contribution for nextcloud/calendar#3919

## TODO
nextcloud/server#41084 needs to remain open as there is a second issue, which prevents the SEQUENCE evaluation in certain cases.

/backport to stable28
/backport to stable27
/backport to stable26

ChristophWurst commented on apps/dav/lib/CalDAV/EventComparisonService.php:
```diff
@@ -65,7 +65,7 @@ private function removeIfUnchanged(VEvent $filterEvent, array &$eventsToFilter):
 			}
 			// events are identical and can be removed
-			if (empty(array_diff($filterEventData, $eventToFilterData))) {
+			if ($filterEventData == $eventToFilterData) {
```
do you have some example values for the two arrays?

Not sure if this breaks the logic but we always use strict comparison:

```suggestion
			if ($filterEventData === $eventToFilterData) {
```

Should be fine here, though, because both vars have the type array.

flortsch replied:
Compared to `==`, the `===` operator also takes the order of array elements into consideration. Not familiar with those array contents and whether Nextcloud normalizes & sorts the data stored in those arrays. If not, I think that this could result in events treated as changed (with notifications being sent out) although the event content itself has not changed and confusing users. So `==` or `array_diff_assoc` would be a more defensive check.

rcwschaller replied:
All three suggested solutions are equally valid here, as the function creates both arrays itself and in the same kind of controlled index sequence. My initial thought was as @flortsch pointed out that == would be more robust with the potential user data, but as @ChristophWurst  pointed out correctly, this is not necessary.

Eventually I also looked at the performance w.r.t. the array_diff_assoc solution:
`php -r '$st = microtime(true); $a1=array("","","",""); $a2=array("","","1",""); for($i = 1; $i <=100000; $i++) { if ($a1==$a2) {} else {} } echo "Execution time = ".(microtime(true)-$st)." sec\n";'
Execution time = 0.0091381072998047 sec`
`php -r '$st = microtime(true); $a1=array("","","",""); $a2=array("","","1",""); for($i = 1; $i <=100000; $i++) { if ($a1===$a2) {} else {} } echo "Execution time = ".(microtime(true)-$st)." sec\n";'
Execution time = 0.0064690113067627 sec`
`php -r '$st = microtime(true); $a1=array("","","",""); $a2=array("","","1",""); for($i = 1; $i <=100000; $i++) { if (empty(array_diff_assoc($a1,$a2))) {} else {} } echo "Execution time = ".(microtime(true)-$st)." sec\n";'
Execution time = 0.012899160385132 sec`

**And so the winner is ===**
[Well, actually performance measurement was as one would expect it to be, but still nice to be confirmed]

miaulalala replied:
Go for it :smile: thanks for the thorough analysis!

tcitworld commented:
Do you mind adding a test to `EventComparisonServiceTest` that demonstrates this behavior is fixed?

rcwschaller commented:
> Do you mind adding a test to `EventComparisonServiceTest` that demonstrates this behavior is fixed?

Done. Not really sure if such a small change warrants the tests, but if it helps the collaboration, I am happy to provide it.

tcitworld reviewed (approved):
Thanks for the effort, appreciated.

Do you mind fixing the lint issues reported by the CI? The other failures are unrelated. We can take care of that if you prefer.

miaulalala commented:
Can you squash all your commits into one?

Easiest is probably:

```
git reset --soft master
git add <your files>
git commit -sm "... your commit message here..."
git push --force-with-lease
```

rcwschaller committed:
fix(dav): wrong comparison method between two events

Old comparison implementation compares each element of the array against each other with no respect for the associated array label, which leads to wrongful removals because one value is accidentally present in a completely different label. New comparison works 'by-label' individually.

Partly fixes #41084 because changes between 'SEQUENCE' not present, 'SEQUENCE:0' and 'SEQUENCE:1' were not detected in the old implementation and thus no email update sent.

Co-authored-by: Christoph Wurst <ChristophWurst@users.noreply.github.com>
Signed-off-by: Robert C. Schaller <gtbc_robert.schaller@rsxc.de>

rcwschaller commented:
@miaulalala should be clean now.

rcwschaller commented:
/backport to stable26

rcwschaller commented:
/backport to stable27

rcwschaller commented:
/backport to stable28
